rest/blip_api_test.go

```diff
@@ -2485,7 +2485,258 @@ func TestBlipDeltaSyncPullRevCache(t *testing.T) {
 	assert.Equal(t, deltaCacheMisses, updatedDeltaCacheMisses)
+}
+
+// Check removal handling for unavailable multi-channel revisions in delta
+// sync when fromRevision is a channel removal.
+func TestBlipDeltaSyncPullRevCacheChannelFromRevRemoval(t *testing.T) {
+	if !base.IsEnterpriseEdition() {
+		t.Skipf("Skipping enterprise-only delta sync test.")
+	}
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+
+	sgUseDeltas := base.IsEnterpriseEdition()
+	rtConfig := RestTesterConfig{
+		DatabaseConfig: &DbConfig{
+			DeltaSync: &DeltaSyncConfig{
+				Enabled: &sgUseDeltas,
+			},
+		},
+	}
+	rt := NewRestTester(t, &rtConfig)
+	defer rt.Close()
+
+	// waitForChangesInChannel asserts that waiting for a number of changes in the specified set of channel(s)
+	// was successful.
+	waitForChangesInChannel := func(channels []string, numChanges int) {
+		changesURL := "/db/_changes?filter=sync_gateway/bychannel&channels=" + strings.Join(channels[:], ",")
+		changes, err := rt.WaitForChanges(numChanges, changesURL, "", true)
+		require.NoError(t, err, "Error retrieving changes results")
+		require.Len(t, changes.Results, numChanges)
+	}
+
+	// Create client with a user who has access to channel1.
+	client1, err := NewBlipTesterClientOptsWithRT(t, rt, &BlipTesterClientOpts{
+		Username:     "alice",
+		Channels:     []string{"channel1"},
+		ClientDeltas: true,
+	})
+	require.NoError(t, err)
+	defer client1.Close()
+
+	// Create client with a user who has access to both channel1 and channel2.
+	client2, err := NewBlipTesterClientOptsWithRT(t, rt, &BlipTesterClientOpts{
+		Username:     "bob",
+		Channels:     []string{"channel1", "channel2"},
+		ClientDeltas: true,
+	})
+	require.NoError(t, err)
+	defer client2.Close()
+
+	// Create the first revision of doc1 on both channel1 and channel2
+	resp := rt.SendAdminRequest(http.MethodPut, "/db/doc1",
+		`{"channels":["channel1","channel2"],"keys":["key1"]}`)
+	require.Equal(t, http.StatusCreated, resp.Code)
+	waitForChangesInChannel([]string{"channel1", "channel2"}, 1)
+
+	// Start one-shot pull at client1 and retrieve the first revision.
+	err = client1.StartOneshotPull()
+	assert.NoError(t, err)
+
+	data, ok := client1.WaitForRev("doc1", "1-ebebfcdd41fb6253d9ab955fa32a9791")
+	require.True(t, ok)
+	assert.Contains(t, string(data), `"channels":["channel1","channel2"]`)
+	assert.Contains(t, string(data), `"keys":["key1"]`)
+
+	// Start one-shot pull at client2 and retrieve the first revision.
+	err = client2.StartOneshotPull()
+	assert.NoError(t, err)
+
+	data, ok = client2.WaitForRev("doc1", "1-ebebfcdd41fb6253d9ab955fa32a9791")
+	require.True(t, ok)
+	assert.Contains(t, string(data), `"channels":["channel1","channel2"]`)
+	assert.Contains(t, string(data), `"keys":["key1"]`)
+
+	// Create the second revision of doc1 on channel2 as removal from channel1.
+	resp = rt.SendAdminRequest(http.MethodPut, "/db/doc1?rev=1-ebebfcdd41fb6253d9ab955fa32a9791",
```

Comment:
I think rev 2 isn't necessary in this test, is it?  Can just be
Client pulls rev 1 (in two channels)
Client pulls rev 2 (in one channel)
Purge rev 2
Client requests rev 3 as delta from rev 2, rev 2 is loaded via removal handling
Verify falls back to full body revision.

Reply:
It is true, as discussed, this test has been replaced with both TestDeltaSyncWhenFromRevIsChannelRemoval and TestDeltaSyncWhenToRevIsChannelRemoval to check the delta sync behavior.